issue 12062 - don't box primitives for some simple cases [ci: last-only]

mkeskells opened the pull request:
undo the wrapping of Predef.double2Double(d).isNaN() to java.lang.Double.isNaN 
undo the wrapping of Predef.float2float(f).isNaN() to java.lang.Float.isNaN 

do the same for other similar methods

- [x] cope with all of the other similar Predef boxing conversions
- [x] cope with similar boxing for calls to `toString`and `hashCode`
- [x] add allocation tests to check for boxing

allocation tests are not great for boxing checks for example in Boolean, but it seemed the simplest way to check for the common cases

fixes https://github.com/scala/bug/issues/12062

mkeskells committed:
don't box primitive values  to call static methods on java.lang.Double/Float/Integer etc
don't box primitives to call hashCode/toString
copy with trees with side effect
add tests to verify no boxing

mkeskells commented:
/rebuild

hrhino commented on src/compiler/scala/tools/nsc/transform/CleanUp.scala:
```diff
@@ -524,4 +565,21 @@ abstract class CleanUp extends Statics with Transform with ast.TreeDSL {
   } // CleanUpTransformer
 
+
+  val doubleAndFloatRedirectMethods = Map(
```
why not `Name`s?

mkeskells replied:
no idea 😞 
fixed now

dwijnand commented on src/reflect/scala/reflect/internal/Definitions.scala:
```diff
@@ -1584,6 +1584,8 @@ trait Definitions extends api.StandardDefinitions {
       lazy val Predef_conforms     = (getMemberIfDefined(PredefModule, nme.conforms)
                                orElse getMemberMethod(PredefModule, TermName("conforms"))) // TODO: predicate on -Xsource:2.10 (for now, needed for transition from M8 -> RC1)
       lazy val Predef_classOf      = getMemberMethod(PredefModule, nme.classOf)
+      lazy val Predef_double2Double= getMemberMethod(PredefModule, nme.double2Double)
```
![](https://i.redd.it/k7syiargegiy.jpg) 😄

dwijnand commented on src/compiler/scala/tools/nsc/transform/CleanUp.scala:
```diff
@@ -524,4 +597,24 @@ abstract class CleanUp extends Statics with Transform with ast.TreeDSL {
+    TermName("intValue") -> TermName("toInt"),
+    TermName("longValue") -> TermName("toLong"),
+    TermName("floatValue") -> TermName("toFloat"),
+    TermName("doubleValue") -> TermName("toDouble")
```
The RHS ones exist as names, as do hashCode and toString (as `hashCode_` and `toString_`).  Should we add the others too?  I'm not sure what all the reasons or principles for using or not using StdNames, but at a guess I think it's better to have them there rather than here.

mkeskells replied:
there was no particular reason for putting then here. Happy to move
presume I should move the `TermName`s, not the Map

dwijnand replied:
Yep - the mapping is specific to the rewrite here.

dwijnand commented on src/compiler/scala/tools/nsc/transform/CleanUp.scala:
```diff
@@ -501,7 +501,80 @@ abstract class CleanUp extends Statics with Transform with ast.TreeDSL {
+          else if (cls == definitions.IntClass) BoxedIntClass
+          else if (cls == definitions.LongClass) BoxedLongClass
+          else if (cls == definitions.FloatClass) BoxedFloatClass
+          else if (cls == definitions.DoubleClass) BoxedDoubleClass
```
Wow, surprisingly it seems this mapping doesn't exist anywhere in the compiler... 🤯

lrytz replied:
Isn't it https://github.com/scala/scala/blob/aeee8f0b2f9f970d07dd33fdb49e93fa1140e103/src/reflect/scala/reflect/internal/Definitions.scala#L99 ?

sjrd commented:
IMO it's worrying that we are now solving an issue with more code in the compiler where a 100% library-based solution would have done the job, and more reliably so:
https://github.com/scala/scala/compare/2.13.x...sjrd:efficient-rich-primitives

Of course, there's the binary compatibility problem now, but still...

dwijnand commented:
> IMO it's worrying that we are now solving an issue with more code in the compiler where a 100% library-based solution would have done the job, and more reliably so:
> [2.13.x...sjrd:efficient-rich-primitives](https://github.com/scala/scala/compare/2.13.x...sjrd:efficient-rich-primitives)
> 
> Of course, there's the binary compatibility problem now, but still...

Wouldn't work, see scala/bug#12062:

> Using `Predef.doubleWrapper(1.0).isNaN` would be preferable because `RichDouble` is a value class. However, `doubleWrapper` is a lower priority implicit than `double2Double` because it's defined in `Predef`'s parent `LowPriorityImplicits`.

dwijnand commented on src/compiler/scala/tools/nsc/transform/CleanUp.scala:
```diff
@@ -524,4 +597,24 @@ abstract class CleanUp extends Statics with Transform with ast.TreeDSL {
+    TermName("intValue") -> termNames.toInt,
+    TermName("longValue") -> termNames.toLong,
+    TermName("floatValue") -> termNames.toFloat,
+    TermName("doubleValue") -> termNames.toDouble
```
The new ones too, IMO.  I think the idea is that it's best to have these all defined once and be in a known location.

lrytz commented on src/compiler/scala/tools/nsc/transform/CleanUp.scala:
```diff
@@ -501,7 +501,80 @@ abstract class CleanUp extends Statics with Transform with ast.TreeDSL {
+      //new Predef.float2Float(x).intValue() -> x.toInt()
+      //new Predef.float2Float(x).longValue() -> x.toLong()
+      //new Predef.float2Float(x).floatValue() -> x.toFloat()
+      //new Predef.float2Float(x).doubleValue() -> x.toDouble()
```
Should we handle conversions from the other primitive types? `1.longValue` has the same issue, it compiles to `int2Integer(1).longValue` (boxes), not `intWrapper(1).longValue` (doesn't box).

On the other hand, is it worth handling these conversions? Are people actually using `xValue` instead of `toX`?

mkeskells replied:
seems easy enough to do. will squash and add this on top to review

mkeskells commented:
@lrytz squashed and updated with your comments
added conversion for the java numbers, seemed easy enough, and would probably get used then people are converting from java code

lrytz commented on src/compiler/scala/tools/nsc/transform/CleanUp.scala:
```diff
@@ -501,6 +501,65 @@ abstract class CleanUp extends Statics with Transform with ast.TreeDSL {
+            val targetMethod = cls.companionModule.info.decl(doubleAndFloatRedirectMethods(methodName))
+            gen.mkMethodCall(targetMethod, params)
+          } else {
+            gen.mkMethodCall(params.head.symbol, javaNumberConversions(methodName), Nil)
```
`mkMethodCall(params.head.symbol, ...)` is incorrect here, it drops the `param` tree.
```suggestion
            gen.mkMethodCall(Select(params.head, javaNumberConversions(methodName)), Nil)
```

lrytz commented on src/compiler/scala/tools/nsc/transform/CleanUp.scala:
```diff
@@ -501,6 +501,65 @@ abstract class CleanUp extends Statics with Transform with ast.TreeDSL {
+      // long2Long
+      // boolean2Boolean
+      //
+      case Apply(Select(Apply(boxing, params), methodName), Nil)
+        if currentRun.runDefinitions.PreDef_primitives2Primitives.contains(boxing.symbol) &&
+          params.size == 1 &&
+          allFloatingMethodsToRewrite.contains(methodName) =>
```
The rewrite drops the qualifier of the boxing method. So for example,

```
scala> {println("hi"); Predef}.int2Integer(1).doubleValue
```

is re-written and the side effect is dropped.
```suggestion
      case Apply(Select(Apply(boxing @ Select(qual, _), params), methodName), Nil)
        if currentRun.runDefinitions.PreDef_primitives2Primitives.contains(boxing.symbol) &&
          params.size == 1 &&
          allFloatingMethodsToRewrite.contains(methodName) &&
          treeInfo.isExprSafeToInline(qual) =>
```

retronym commented on src/compiler/scala/tools/nsc/transform/CleanUp.scala:
```diff
@@ -524,4 +584,24 @@ abstract class CleanUp extends Statics with Transform with ast.TreeDSL {
   } // CleanUpTransformer
 
+
+  val BOX = termNames.box
```
Could these vals be private?

mkeskells replied:
yes - done

retronym commented on src/compiler/scala/tools/nsc/transform/CleanUp.scala:
```diff
@@ -524,4 +584,24 @@ abstract class CleanUp extends Statics with Transform with ast.TreeDSL {
   } // CleanUpTransformer
 
+
+  val BOX = termNames.box
```
Within the compiler, we use `nme` instead of the longer alias `termNames`.

mkeskells replied:
done

lrytz commented on src/compiler/scala/tools/nsc/transform/CleanUp.scala:
```diff
@@ -501,6 +501,66 @@ abstract class CleanUp extends Statics with Transform with ast.TreeDSL {
+      // similarly for all primitive types
+      case Apply(Select(Apply(box @ Select(boxer, _), params), methodName), Nil)
+        if objectMethods.contains(methodName) &&
+          params.size == 1
+          && currentRun.runDefinitions.isBox(box.symbol)
```
In principle this would be more correct

```suggestion
          params.size == 1 &&
          currentRun.runDefinitions.isBox(box.symbol) &&
          treeInfo.isExprSafeToInline(boxer)
```

Even if in reality that the backend currently drops the prefix; this is a bug that should be fixed in the backend.

```
scala> {println("hi"); Int}.box(1).hashCode
val res1: Int = 1
```

lrytz commented:
(rebased on current 2.12.x)

lrytz commented on test/files/run/t12062.scala:
```diff
@@ -0,0 +1,124 @@
+    floating("TestFloat", "val value:Float  = 1.toFloat"),
+    floating("TestDouble", "val value:Double  = 1.toDouble")
+    )
+  def primatives(className: String, decl: String) = {
```
🦍

lrytz reviewed (approved):
LGTM. Pushed a tiny cleanup. We should squash everything before merging.
